File: src/main/java/core/scm/specialforms/New.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.evaluator.Reflector;
import core.exceptions.IllegalSyntaxException;

import java.util.List;

public enum New implements ISpecialForm {
  NEW;

  private final Reflector reflector = new Reflector();

  @Override
  public Object eval(List<Object> expression, Environment env, Evaluator evaluator) {
    if (expression.size() < 2) {
      throw IllegalSyntaxException.of(toString(), expression);
    }
    String clazz = expression.get(1).toString();

    Object[] args = new Object[expression.size() - 2];
    for (int i = 0; i < args.length; i++) {
      args[i] = evaluator.eval(expression.get(i + 2), env);
    }
    return reflector.newInstance(clazz, args);
  }

  @Override
  public String toString() {
    return "new";
  }
}

File: src/main/java/core/scm/specialforms/New.kt
package core.scm.specialforms

import core.Evaluator
import core.Reflector
import core.exceptions.IllegalSyntaxException
import core.Writer

object New : SpecialForm("new") {

    private val reflector = Reflector()

    override fun eval(form: List<Any?>, evaluator: Evaluator): Any {
        if (form.size < 2) {
            throw IllegalSyntaxException(toString(), Writer.write(form))
        }
        val constructorArgs = arrayOfNulls<Any>(form.size - 2)
        constructorArgs.indices.forEach { i -> constructorArgs[i] = evaluator.eval(form[i + 2]) }
        val clazz = form[1].toString()
        return reflector.newInstance(clazz, constructorArgs)
    }
}
